teps/0032-tekton-notifications.md

```diff
@@ -0,0 +1,174 @@
+- Write a controller for cloud events (even if that could be part of the solution).
+  See [tektoncd/pipeline#2944](https://github.com/tektoncd/pipeline/issues/2944) for
+  further context.
+
+### Use Cases (optional)
+
+- Notify individuals or groups about the failure of a pipeline
+- Notify individuals or groups about the completion of a long running task
+- Notify individuals or groups about the start or a pipeline that was triggered as
+  a consequence of an automated event (cron, pull request, manual approval)
+
+## Requirements
+
+***This section is a stub and subject to change.***
+
+Once notifications are implemented, we do not want to force Tekton users to modify
+all their existing tasks or pipelines to enable notifications for them:
+
+- Allow using catalog tasks as is
+- Allow notifications for pipelines with no change to the pipeline itself
+
+For installations that run large number of task and pipelines, the configuration
+of notification should be designed to be scalable and easy to maintain.
+
+- Allow selecting pipelines and tasks to be notified about via selectors
+- Allow selecting events to be notified about
+- Allow pipeline and task authors to re-use of notification "rules"
+
+## Proposal
+
+***This section is a stub and subject to change.***
+
+The proposal is to introduce notification policies and templates.
+
+Notification policies define condition under which we want to trigger a notification;
+they would allow users to:
+
+- filter sources such using rules (using CEL?)
+  - kind: task, pipeline
+  - metadata: name, labels, namespace, annotations(?)
+- filter events such as start, run, success, fail (using CEL or syntax sugar on top)
+
+Notification templates define how a notification is sent. They would allow user to:
+
+- define the task / pipeline to be executed
+- bind an event type with the inputs to the task / param
+
+Both notification policies and templates could be defined a dedicated CRD.
+The policy controller would be responsible to provision `trigger` resources by
+combining the information from the policy and the templates.
+
+Which notification template is triggered is decided based on runtime information passed to the `TaskRun`
+or `PipelineRun`, or can be configured on Tekton for each namespace and added to resources
+as an annotation.
+
+All resources generated or associated with a notification policy and a notification
+template will be automatically labelled by the notification controller so that they
+can be easily identified and filtered.
+
+### User Experience (optional)
+
+Notification targeted to humans is the main goal of this TEP because I would like to make
+sure that requirements for machine targeted notification do not pollute the user experience
+of tekton native notifications.
+
+User experience is very important for this proposal. Notification should be easy to use
+and maintain across large number of tasks and pipelines.
+
+## Alternatives
```

Comment:
How do other eventing projects like Knative play into this proposal? Is there any existing work we can leverage here, and what advantages do we get by introducing our own concepts?

Reply:
I'm not aware of any Knative specific concept / CRD that would help on this specifically.

Knative eventing provides an event bus, so it would be of course possible to inject Tekton generated cloud events into a knative bus and having them delivered based on knative based rules (channels, filters etc). In fact I think cloud events would still be one of the formats notifications can "materialise" into. 

Sending cloud events and piping into knative is possible today, and while it gives users the full flexibility of building an asynchronous workflow anyways they like, it does not specifically provide a notification mechanism for Tekton.

I would like to avoid building a feature in Tekton that relies on any of the knative controllers, as that would very much grow the footprint. I'm definitely in favour of building with interoperability in mind.